File: apps/api/src/schemas.py
"""
Pydantic schemas for the Math Tutor API.

These schemas mirror the TypeScript types in @math-tutor/core for consistency.
"""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class TutoringMode(str, Enum):
    """Available tutoring modes."""

    HINT = "hint"
    CHECK_STEP = "check_step"
    EXPLAIN = "explain"


class GradeLevel(str, Enum):
    """Ontario K-12 grade levels."""

    K = "K"
    G1 = "1"
    G2 = "2"
    G3 = "3"
    G4 = "4"
    G5 = "5"
    G6 = "6"
    G7 = "7"
    G8 = "8"
    G9 = "9"
    G10 = "10"
    G11 = "11"
    G12 = "12"


class ChatRequest(BaseModel):
    """Request schema for the /chat endpoint."""

    question: str = Field(..., min_length=1, description="The math question or problem")
    attempt: Optional[str] = Field(None, description="Student's attempt at the problem")
    mode: TutoringMode = Field(..., description="Tutoring mode (hint, check_step, explain)")
    grade: GradeLevel = Field(..., description="Student's grade level")
    dont_reveal_answer: bool = Field(True, description="Whether to avoid revealing the final answer")
    topic_tags: Optional[list[str]] = Field(None, description="Optional topic tags for context")

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "question": "What is 2 + 3?",
                    "mode": "hint",
                    "grade": "1",
                    "dont_reveal_answer": True,
                }
            ]
        }
    }


class DebugInfo(BaseModel):
    """Debug information included in responses."""

    selected_policy: str = Field(..., description="The policy that was applied")


class ChatResponse(BaseModel):
    """Response schema for the /chat endpoint."""

    response: str = Field(..., description="The tutor's response")
    refusal: bool = Field(..., description="Whether the request was refused (off-topic)")
    citations: Optional[list[str]] = Field(None, description="Any citations or references")
    debug: Optional[DebugInfo] = Field(None, description="Debug information")

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "response": "Great question! Let's think about this together. What do you get when you count up 2 from 3?",
                    "refusal": False,
                    "debug": {"selected_policy": "hint_mode"},
                }
            ]
        }
    }

File: packages/core/src/schemas.ts
/**
 * @math-tutor/core - Shared schemas and types
 *
 * These schemas define the contract between frontend, backend, and evaluation.
 * Use Zod for runtime validation and TypeScript inference.
 */

import { z } from "zod";

// =============================================================================
// Tutoring Mode
// =============================================================================

export const TutoringMode = z.enum(["hint", "check_step", "explain"]);
export type TutoringMode = z.infer<typeof TutoringMode>;

export const TUTORING_MODE_DESCRIPTIONS: Record<TutoringMode, string> = {
  hint: "Provide guiding questions to help discover the answer",
  check_step: "Validate student's work and suggest next steps",
  explain: "Explain the concept without revealing the answer",
};

// =============================================================================
// Grade Level
// =============================================================================

export const GradeLevel = z.enum([
  "K",
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
]);
export type GradeLevel = z.infer<typeof GradeLevel>;

// =============================================================================
// Chat Request/Response
// =============================================================================

export const ChatRequest = z.object({
  question: z.string().min(1, "Question is required"),
  attempt: z.string().optional(),
  mode: TutoringMode,
  grade: GradeLevel,
  dont_reveal_answer: z.boolean().default(true),
  topic_tags: z.array(z.string()).optional(),
});
export type ChatRequest = z.infer<typeof ChatRequest>;

export const ChatResponse = z.object({
  response: z.string(),
  refusal: z.boolean(),
  citations: z.array(z.string()).optional(),
  debug: z
    .object({
      selected_policy: z.string(),
    })
    .optional(),
});
export type ChatResponse = z.infer<typeof ChatResponse>;

// =============================================================================
// Chat Message (for UI state)
// =============================================================================

export const MessageRole = z.enum(["user", "assistant", "system"]);
export type MessageRole = z.infer<typeof MessageRole>;

export const ChatMessage = z.object({
  id: z.string(),
  role: MessageRole,
  content: z.string(),
  timestamp: z.number(),
  metadata: z
    .object({
      mode: TutoringMode.optional(),
      grade: GradeLevel.optional(),
      refusal: z.boolean().optional(),
    })
    .optional(),
});
export type ChatMessage = z.infer<typeof ChatMessage>;

// =============================================================================
// Evaluation Rubric
// =============================================================================

export const TutorRubricResult = z.object({
  no_answer_reveal: z.boolean(),
  socratic_step: z.boolean(),
  correct_direction: z.boolean().nullable(), // null = not evaluated
  on_topic: z.boolean(),
});
export type TutorRubricResult = z.infer<typeof TutorRubricResult>;

export const EvalResult = z.object({
  id: z.string(),
  input: z.string(),
  expected: z.string().optional(),
  actual: z.string(),
  metrics: z.record(z.union([z.number(), z.boolean(), z.null()])),
});
export type EvalResult = z.infer<typeof EvalResult>;

// =============================================================================
// Training Data Schema
// =============================================================================

export const TrainingExample = z.object({
  id: z.string(),
  grade: GradeLevel,
  topic_tags: z.array(z.string()),
  prompt: z.string(),
  response: z.string(),
  metadata: z.object({
    source: z.string(),
    difficulty: z.enum(["easy", "medium", "hard"]).optional(),
    mode: TutoringMode.optional(),
  }),
});
export type TrainingExample = z.infer<typeof TrainingExample>;

